sh.c

```diff
@@ -1501,7 +1501,7 @@ text comp_putchar_inline(ast param) {
 
   if (contains_side_effects) {
     ident = fresh_ident();
-    append_glo_decl(string_concat3(comp_lvalue(ident), wrap_char('='), res));
+    append_glo_decl(string_concat4(comp_lvalue(ident), wrap_str("=$(("), res, wrap_str("))")));
```

Comment:
When the expression doesn't contain a side effect, it's placed in an arithmetic expansion, but when it does have side effects, we didn't wrap it in `$(( ))` which caused statements such as `putchar(i++)` to fail.